### PR TITLE
tests/*: introduce go-dmflakey

### PR DESCRIPTION
**Pull request by fuweid**
This commit is to add go-binding for go-dmflakey. It's used to simulate powerfailure with common filesystems.

> NOTE: If it's acceptable, I will mark my original repo https://github.com/fuweid/go-dmflakey as read-only and add comment that it has been moved to bbolt repo.

### CODE REVIEW

File: .github/workflows/failpoint_test.yaml [hunk 1]
```diff
@@ -15,5 +15,6 @@ jobs:
         with:
           go-version: ${{ steps.goversion.outputs.goversion }}
       - run: |
+          sudo make root-test
```
Comment (ahrtr) [inline .github/workflows/failpoint_test.yaml:18]: Suggest to add a separate workflow test for the data corruption test, i.e. `data_corruption_test.yaml`. Eventually we will setup nightly test for it.

We can address this in a followup PR along with https://github.com/etcd-io/bbolt/pull/622

File: tests/dmflakey/loopback.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	return "", fmt.Errorf("failed to associate free loop device with backing file %s after retry %v",
+		backingFile, maxRetryToAttach)
+}
+
+// detachLoopDevice disassociates the loop device from any backing file.
+//
+// REF: https://man7.org/linux/man-pages/man4/loop.4.html
+func detachLoopDevice(loopDevice string) error {
+	loopFd, err := os.Open(loopDevice)
```
Comment (ahrtr) [inline tests/dmflakey/loopback.go:67]: `os.Open` opens the file in `O_RDONLY` mode. When attaching the device, you open the file in `os.O_RDWR` mode. Please double check this.

**Comment by fuweid** [comment]
ping @ahrtr @serathius @jmhbnz 

REF: https://github.com/etcd-io/bbolt/pull/622#discussion_r1413073448

File: tests/dmflakey/dmsetup.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"golang.org/x/sys/unix"
+)
+
+// newFlakeyDevice creates flakey device.
+//
+// REF: https://docs.kernel.org/admin-guide/device-mapper/dm-flakey.html
+func newFlakeyDevice(flakeyDevice, loopDevice string, interval time.Duration) error {
+	loopSize, err := getBlkSize(loopDevice)
+	if err != nil {
+		return fmt.Errorf("failed to get loop device %s size: %w", loopDevice, err)
```
Comment (ahrtr) [inline tests/dmflakey/dmsetup.go:21]:
```suggestion
		return fmt.Errorf("failed to get the size of the loop device %s: %w", loopDevice, err)
```

File: tests/dmflakey/dmflakey.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+)
+
+type featCfg struct {
+	// SyncFS attempts to synchronize filesystem before inject failure.
+	syncFS bool
+	// interval is used to determine how long the failure lasts.
+	interval time.Duration
+}
+
+var defaultFeatCfg = featCfg{interval: defaultInterval}
```
Comment (ahrtr) [inline tests/dmflakey/dmflakey.go:24]: Where is `defaultInterval`? Please define all constants right above this line.

File: tests/dmflakey/dmsetup.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	// The flakey device will be available in interval.Seconds().
+	table := fmt.Sprintf("0 %d flakey %s 0 %d 0",
```
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/dmflakey/dmsetup.go:25]: Updated. Please take a look.

File: tests/dmflakey/dmflakey.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+	// SyncFS attempts to synchronize filesystem before inject failure.
+	syncFS bool
+	// interval is used to determine how long the failure lasts.
+	interval time.Duration
+}
+
+// Default values.
+var (
+	defaultImgSize  int64 = 1024 * 1024 * 1024 * 10 // 10 GiB
+	defaultInterval       = 2 * time.Minute
```
Comment (ahrtr) [inline tests/dmflakey/dmflakey.go:27]: Please define the two items as constants.

File: tests/dmflakey/dmflakey.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+type featCfg struct {
+	// SyncFS attempts to synchronize filesystem before inject failure.
+	syncFS bool
+	// interval is used to determine how long the failure lasts.
+	interval time.Duration
```
[the remark this answers is not in the document]
Comment (fuweid) [reply tests/dmflakey/dmflakey.go:25]: Yes. Updated.

File: tests/dmflakey/dmflakey.go (hunk 6 above)
Comment (fuweid) [reply tests/dmflakey/dmflakey.go]: Sure. Updated

**Pull request merged**